lucene/codecs/src/java/org/apache/lucene/codecs/uniformsplit/IntersectBlockReader.java

```diff
@@ -0,0 +1,315 @@
+        if (runAutomatonFromPrefix(lineTerm)) {
+          return lineTerm;
+        } else if (beyondCommonPrefix) {
+          return null;
+        }
+      }
+
+    } while (true);
+  }
+
+  protected boolean runAutomatonFromPrefix(BytesRef term) {
+    int state = runAutomatonForState(term.bytes, term.offset + blockPrefixLen, term.length - blockPrefixLen, blockPrefixRunAutomatonState);
+    if (state >= 0 && runAutomaton.isAccept(state)) {
+      return true;
+    }
+    if (isBeyondCommonPrefix(term)) {
+      // If the automaton rejects early the term, before the common prefix length,
+      // and if the term rejected byte is lexicographically after the same byte in the common prefix,
+      // then it means the current term is beyond the common prefix.
+      // Exit immediately the enumeration.
+      beyondCommonPrefix = true;
+    }
+    return false;
+  }
+
+  /**
+   * Run the automaton and return the final state (not necessary accepted). -1 signifies no state / no match.
+   * Sets {@link #numBytesAccepted} with the offset of the first byte rejected by the automaton;
+   * or (offset + length) if no byte is rejected.
+   */
+  protected int runAutomatonForState(byte[] s, int offset, int length, int initialState) {
+    //see ByteRunAutomaton.run(); similar
+    int state = initialState;
+    int index = 0;
+    while (index < length) {
+      state = runAutomaton.step(state, s[index + offset] & 0xFF);
+      if (state == -1) {
+        break;
+      }
+      index++;
+    }
+    numBytesAccepted = index;
+    return state;
+  }
+
+  /**
+   * Determines if the provided {@link BytesRef} is beyond the automaton common prefix.
+   * This method must be called after a call to {@link #runAutomatonForState} because
+   * it uses {@link #numBytesAccepted} value.
+   */
+  protected boolean isBeyondCommonPrefix(BytesRef bytesRef) {
+    // If the automaton rejects early the bytes, before the common prefix length,
+    // and if the rejected byte is lexicographically after the same byte in the common prefix,
+    // then it means the bytes are beyond the common prefix.
+    return numBytesAccepted < commonPrefixRef.length
+        && numBytesAccepted < bytesRef.length
+        && (bytesRef.bytes[numBytesAccepted + bytesRef.offset] & 0xFF) > (commonPrefixRef.bytes[numBytesAccepted + commonPrefixRef.offset] & 0xFF);
+  }
+
+  @Override
+  public boolean seekExact(BytesRef text) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void seekExact(long ord) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public SeekStatus seekCeil(BytesRef text) {
+    throw new UnsupportedOperationException();
+  }
+
+  @Override
+  public void seekExact(BytesRef term, TermState state) {
+    throw new UnsupportedOperationException();
+  }
+
+  //TODO this approach is a hack to reuse complex code in AutomatonTermsEnum.  How to make ATE's logic more reusable?
```

Comment:
@rmuir this is the hack.  The conundrum I ask in the comment:  "How to make ATE's logic more reusable"?  Your last comment in JIRA seemed to indicate you didn't think AutomatonTermsEnum should be modified at all.  The changes to ATE in this patch are trivial and are avoidable if we choose not to care much about "isLinearState" -- fine by me.  Nonetheless the hack is here (IntersectBlockReader) not ATE, in how ATE is subclassed just to call its logic.  I'm not sure if you've seen the code yet.

Reply:
Any opinion @jpountz ?

Reply:
Maybe fork the part that you need from AutomatonTermsEnum instead?

Reply:
Okay.  Trade-offs... I usually prefer to re-use code instead of duplicating code.  This would duplicate a lot of complexities in ATE; and you find that better?

Reply:
Yes I like it a bit better, it avoids exposing internals of ATE and makes it easier to evolve both features independently.